### PR TITLE
Add fn demo command

### PR DESCRIPTION
**Pull request by wenchajun**
Add fn demo command

Signed-off-by: chengdehao <dehaocheng@yunify.com>

### CODE REVIEW

**Comment by benjaminhuo** [comment]
The binary shouldn't be put into the source repo maybe?
We can launch a release for cli to publish binaries

File: docs/demo.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+## Parameters
+
+```shell
+--dry-run          Used to prompt for the components and their versions to be installed by the current command.
+--region-cn        For users in China to speed up the download process of dependent components.
+--delete           Remove the demo environment
+--verbose          Show verbose information.
+--version string   Used to specify the version of OpenFunction to be installed. The permitted versions are: v0.3.1, v0.4.0, latest. (default "v0.4.0")
+```
```
Comment (tpiperatgod) [inline docs/demo.md]: can pass the `--timeout` option now

File: pkg/cmd/subcommand/demo.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+	if i.DryRun {
+		return nil
+	}
+	if !continueFunc() {
+		return nil
+	}
+
+	grp, gctx := errgroup.WithContext(ctx)
+	defer gctx.Done()
```
Comment (tpiperatgod) [inline pkg/cmd/subcommand/demo.go]: need to fix the defer bug like #9

File: pkg/cmd/subcommand/demo.go [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+func (i *Demo) RunDemo(l *k8s.Clientset, cmd *cobra.Command) error {
+	operator := common.NewOperator(runtime.GOOS, i.OpenFunctionVersion, i.RegionCN, i.Verbose)
+	ti := util.NewTaskInformer("RunDemo")
+	start := time.Now()
+	ctx, done := context.WithCancel(
+		context.Background(),
+	)
+	defer done()
+	grp, gctx := errgroup.WithContext(ctx)
+	defer gctx.Done()
```
Comment (tpiperatgod) [inline pkg/cmd/subcommand/demo.go]: need to fix the defer bug like #9

File: pkg/dependency/common/common.go [hunk 4]
```diff
@@ -486,6 +488,49 @@ func (o *Operator) UninstallOpenFunction(ctx context.Context, cl *k8s.Clientset,
 	return nil
 }
 
+func (o *Operator) DownloadKind(ctx context.Context) error {
+	DownloadCmd := "curl -Lo ./kind https://kind.sigs.k8s.io/dl/v0.11.1/kind-linux-amd64"
```
Comment (tpiperatgod) [inline pkg/dependency/common/common.go]: `curl` is not a generic command and it should be in the linux implementation.

**Review by tpiperatgod — CHANGES_REQUESTED** [review]

File: pkg/cmd/subcommand/demo.go (hunk 2 above)
Comment (benjaminhuo) [reply pkg/cmd/subcommand/demo.go]: @wenchajun any update?

File: pkg/cmd/subcommand/demo.go (hunk 3 above)
Comment (benjaminhuo) [reply pkg/cmd/subcommand/demo.go]: any update?

**Comment by benjaminhuo** [comment]
It's ready to review now? Please solve the conflict first, thanks @wenchajun

File: docs/demo.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
+# fn demo
```
Comment (benjaminhuo) [inline docs/demo.md]: fn -> ofn

File: docs/demo.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+This command will help you to create a kind cluster to install OpenFunction and its demo.
+
+## Parameters
+
+```shell
+--dry-run          Used to prompt for the components and their versions to be installed by the current command.
+--region-cn        For users in China to speed up the download process of dependent components.
+--delete           Remove the demo environment
+--verbose          Show verbose information.
+--timeout duration   Set timeout time. Default is 5 minutes. (default 5m0s)
```
Comment (benjaminhuo) [inline docs/demo.md]: duration can be removed?

File: docs/demo.md [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+#### Remove the demo environment
+
+```shell
+fn demo --delete
+```
+
+
+
+#### Support for users in China to speed up the installation process
```
Comment (benjaminhuo) [inline docs/demo.md]: Support for users in China to speed up the installation process
=>
For users who have limited access to gcr.io or github.com

File: docs/demo.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
+# fn demo
+
+This command will help you to create a kind cluster to install OpenFunction and its demo.
+
+## Parameters
+
+```shell
+--dry-run          Used to prompt for the components and their versions to be installed by the current command.
+--region-cn        For users in China to speed up the download process of dependent components.
```
Comment (benjaminhuo) [inline docs/demo.md]: Install OpenFunction with limited access to gcr.io or github.com

File: docs/demo.md [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+```
+
+
+
+#### Support for users in China to speed up the installation process
+
+> This requires that you should also add the `--region-cn` parameter when executing the uninstall operation
+
+```shell
+fn demo --region-cn --all
```
Comment (benjaminhuo) [inline docs/demo.md]: why adding `--all` flag to demo? It should install all components by default

File: pkg/cmd/subcommand/demo.go [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+	var cl *k8s.Clientset
+
+	i := NewDemo(ioStreams)
+
+	cmd := &cobra.Command{
+		Use:                   "demo <args>...",
+		DisableFlagsInUseLine: true,
+		Short:                 "Create OpenFunction demo.",
+		Long: `
+This command will help you to Run the OpenFunction demo.
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/demo.go]: The description could be shorter:
```
You can use ofn demo to run the OpenFunction demo.
Available options are: --delete, --region-cn, --version, --verbose, --dry-run
```

or put the options in `ofn demo -h/--help`
```
You can use ofn demo to run the OpenFunction demo.
```

File: pkg/cmd/subcommand/demo.go [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+					fmt.Sprintf("--version=latest")},
+				)
+				installCMD.Execute()
+				util.CheckErr(i.RunDemo(cl, cmd))
+			}
+		},
+	}
+
+	cmd.Flags().BoolVar(&i.Verbose, "verbose", false, "Show verbose information.")
+	cmd.Flags().BoolVar(&i.RegionCN, "region-cn", false, "For users in China to uninstall dependent components.")
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/demo.go]: For users who have limited access to gcr.io or github.com

File: pkg/cmd/subcommand/demo.go [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+	cmd.Flags().BoolVar(&i.DryRun, "dry-run", false, "Used to prompt for the components and their versions to be uninstalled by the current command.")
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/demo.go]: this is for demo, not uninstall.
And do we need a dry-run for demo? @wenchajun @tpiperatgod

File: pkg/cmd/subcommand/demo.go [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+	return cmd
+}
+
+
+func (i *Demo) RunKind(cl *k8s.Clientset, cmd *cobra.Command) error {
+	operator := common.NewOperator(runtime.GOOS, i.OpenFunctionVersion, i.Timeout, i.RegionCN, i.Verbose)
+	ti := util.NewTaskInformer("Demo")
+	continueFunc := func() bool {
+		reader := bufio.NewReader(os.Stdin)
+		fmt.Fprintln(w, ti.BeforeTask("We will create a Kind cluster and run the OpenFunction Demo..\n"+
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/demo.go]: We will create a Kind cluster and run the OpenFunction Demo..
=>
A Kind cluster will be created and the OpenFunction Demo will be launched in it.

File: pkg/cmd/subcommand/demo.go [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+			"Make sure you know what happens when you do this.\n"+
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/demo.go]: `Make sure you know what happens when you do this.` can be removed

File: pkg/cmd/subcommand/demo.go [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+
+	ctx, done := context.WithTimeout(
+		context.Background(),
+		i.Timeout,
+	)
+	defer done()
+
+	inventoryPending := i.checkConditionsAndGetInventory()
+
+	fmt.Fprintln(w, ti.BeforeTask("Start to run OpenFunction demo.\n"+
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/demo.go]: Start to run OpenFunction demo.Here are the components and corresponding versions to be ran for this installation:
=> 
Launching OpenFunction demo...
The following components will be installed for this demo:

File: pkg/cmd/subcommand/demo.go [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+	cmd.Flags().MarkHidden("keda")
+	cmd.Flags().MarkHidden("dapr")
+	return cmd
+}
+
+
+func (i *Demo) RunKind(cl *k8s.Clientset, cmd *cobra.Command) error {
+	operator := common.NewOperator(runtime.GOOS, i.OpenFunctionVersion, i.Timeout, i.RegionCN, i.Verbose)
+	ti := util.NewTaskInformer("Demo")
+	continueFunc := func() bool {
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/demo.go]: continueFunc => continue
Simply continue is ok

File: pkg/cmd/subcommand/demo.go [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+	RegionCN            bool
+	OpenFunctionVersion string
+	DryRun              bool
+	Delete              bool
+	Timeout             time.Duration
+}
+
+func init() {
+	w = os.Stdout
+	availableVersions = []string{"v0.3.1", "v0.4.0", "latest"}
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/demo.go]: Need to figure out a better way instead of hardcoding supported versions

File: pkg/cmd/subcommand/demo.go [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+				return false
+			}
+		}
+	}
+
+	ctx, done := context.WithTimeout(
+		context.Background(),
+		i.Timeout,
+	)
+	defer done()
```
Comment (tpiperatgod) [inline pkg/cmd/subcommand/demo.go:105]: need to delete kind cluster in the defer func

File: pkg/components/common/common.go [hunk 19]
```diff
@@ -282,6 +290,107 @@ func (o *Operator) Uninstall(
 	return nil
 }
 
+func (o *Operator) DownloadKind(ctx context.Context) error {
+	return o.executor.DownloadKind()
+}
+
+func (o *Operator) CreateKindCluster(ctx context.Context) error {
+	cmd := "kind create cluster"
```
Comment (tpiperatgod) [inline pkg/components/common/common.go]: Can we add the `--name` and `--image` options?
Like `kind create cluster --name openfunction --image kindest/node:v1.20.0`
For OpenFunction v0.4.0, maybe we can specify the Kubernetes of v1.20.0 by default?

File: pkg/components/common/common.go [hunk 20]
```diff
@@ -282,6 +290,107 @@ func (o *Operator) Uninstall(
...
+	if _, _, err := o.executor.Exec(cmd); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (o *Operator) DeleteKind(ctx context.Context) error {
+	cmd := "kind delete cluster"
```
Comment (tpiperatgod) [inline pkg/components/common/common.go]: If we agree to use `--name` to create the Kind cluster, then we also need to add the `--name` option when deleting it

File: pkg/components/linux/linux.go [hunk 21]
```diff
@@ -168,3 +174,28 @@ func (e *Executor) GetInventoryRecord(ctx context.Context) (*inventory.Record, e
...
+func (e *Executor) DownloadKind() error {
+	DownloadCmd := "curl -Lo ./kind https://kind.sigs.k8s.io/dl/v0.11.1/kind-linux-amd64"
+	if _, _, err := e.Exec(DownloadCmd); err != nil {
+		return err
+	}
+	ChmodCmd := "chmod +x ./kind"
+	if _, _, err := e.Exec(ChmodCmd); err != nil {
+		return err
+	}
+	MvCmd := "mv ./kind /usr/bin/kind"
```
Comment (tpiperatgod) [inline pkg/components/linux/linux.go]: `/usr/local/bin/kind` would be better

**Review by tpiperatgod — CHANGES_REQUESTED** [review]

File: pkg/cmd/subcommand/demo.go (hunk 18 above)
Comment (benjaminhuo) [reply pkg/cmd/subcommand/demo.go:105]: Maybe should delete the kind cluster after running RunDemo, not after installing kind by `RunKind`.
If we add the kind cluster auto-deleting/release mechanism, maybe the `--delete` option is not needed.
A user just need to use run `ofn demo xxx` to figure out how of works, and needn't care about the demo env setup or cleanup?

Comment (tpiperatgod) [reply pkg/cmd/subcommand/demo.go:105]: Yes, we have found the right place to insert the logic for deleting the kind of cluster that failed to be created.
We should perform a rollback operation (i.e. delete similar clusters) after an installation failure of OpenFunction to ensure that everything is back to normal the next time the user executes the `ofn demo` command.

**Comment by benjaminhuo** [comment]
@wenchajun would you make corresponding changes to the version list similar to https://github.com/OpenFunction/cli/pull/14 ?

**Comment by wenchajun** [comment]
> @wenchajun would you make corresponding changes to the version list similar to #14 ?
  
Yes，We had a discussion and I was going to make the demo components independent to get a lower coupling. I will modify this demo architecture and update it.

File: docs/demo.md [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+--verbose          Show verbose information.
+```
+
+## Use Cases
+
+
+
+#### For users who have limited access to gcr.io or github.com
+
+> This requires that you should also add the `--region-cn` parameter when executing the uninstall operation
```
Comment (tpiperatgod) [inline docs/demo.md]: This line should be removed

File: pkg/cmd/subcommand/demo.go [hunk 23]
```diff
@@ -0,0 +1,490 @@
...
+	cmd.Flags().DurationVar(&i.Timeout, "timeout", 10*time.Minute, "Set timeout time. Default is 10 minutes.")
+	// In order to avoid too many options causing misunderstandings among users,
+	// we have hidden the following parameters,
+	// but you can still find their usage instructions in the documentation.
+	cmd.Flags().MarkHidden("version")
+	cmd.Flags().MarkHidden("ingress")
+	cmd.Flags().MarkHidden("cert-manager")
+	cmd.Flags().MarkHidden("shipwright")
+	cmd.Flags().MarkHidden("keda")
+	cmd.Flags().MarkHidden("dapr")
```
Comment (tpiperatgod) [inline pkg/cmd/subcommand/demo.go]: seems no need to hide these options

File: pkg/cmd/subcommand/demo.go [hunk 24]
```diff
@@ -0,0 +1,490 @@
...
+	return cmd
+}
+
+func (i *Demo) RunKind(cl *k8s.Clientset, cmd *cobra.Command) error {
+	if i.AutoPrune {
+		defer i.DeleteKind()
+	}
+
+	operator := common.NewOperator(runtime.GOOS, i.OpenFunctionVersion, i.Timeout, i.RegionCN, i.Verbose)
+	ti := util.NewTaskInformer("Demo")
```
Comment (tpiperatgod) [inline pkg/cmd/subcommand/demo.go]: Demo -> DEMO

better to use the uppercase name to keep uniform

File: pkg/cmd/subcommand/demo.go [hunk 25]
```diff
@@ -0,0 +1,490 @@
...
+	Continue := func() bool {
```
Comment (tpiperatgod) [inline pkg/cmd/subcommand/demo.go:81]: seems no need to wait for confirmation?

File: pkg/cmd/subcommand/demo.go [hunk 26]
```diff
@@ -0,0 +1,490 @@
...
+	defer done()
+	inventoryPending := i.checkConditionsAndGetInventory()
+
+	fmt.Fprintln(w, ti.BeforeTask("Launching OpenFunction demo...\n"+
+		"The following components will be installed for this demo:"))
+	printInventory(inventoryPending)
+
+	if i.DryRun {
+		return nil
+	}
```
Comment (tpiperatgod) [inline pkg/cmd/subcommand/demo.go]: can be removed

File: pkg/cmd/subcommand/demo.go [hunk 27]
```diff
@@ -0,0 +1,490 @@
...
+		"OpenFunction": i.OpenFunctionVersion,
+	}
+
+	return inventory
+}
+
+func (i *Demo) InstallKind(ctx context.Context, operator *common.Operator) error {
+	ctx, done := context.WithCancel(ctx)
+	defer done()
+	ti := util.NewTaskInformer("Kind")
```
Comment (tpiperatgod) [inline pkg/cmd/subcommand/demo.go]: Kind -> KIND

better to use the uppercase name to keep uniform

File: pkg/cmd/subcommand/demo.go [hunk 28]
```diff
@@ -0,0 +1,490 @@
...
+
+	fmt.Fprintln(w, ti.TaskInfo("Installing Kind..."))
+	fmt.Fprintln(w, ti.TaskInfo("Downloading Kind binary..."))
+	if err := operator.DownloadKind(ctx); err != nil {
+		return errors.Wrap(err, ti.TaskFailWithTitle("Failed to download Kind"))
+	}
+	fmt.Fprintln(w, ti.TaskInfo("Initializing Kind with Kubernetes mode..."))
```
Comment (tpiperatgod) [inline pkg/cmd/subcommand/demo.go]: "Initializing Kind with Kubernetes mode..." -> "Creating cluster..." or something like this

File: pkg/cmd/subcommand/demo.go [hunk 29]
```diff
@@ -0,0 +1,490 @@
...
+}
+
+func (i *Demo) DeleteKind() error {
+	operator := common.NewOperator(runtime.GOOS, i.OpenFunctionVersion, i.Timeout, i.RegionCN, i.Verbose)
+	ctx, done := context.WithTimeout(
+		context.Background(),
+		i.Timeout,
+	)
+	defer done()
+	ti := util.NewTaskInformer("DeleteKind")
```
Comment (tpiperatgod) [inline pkg/cmd/subcommand/demo.go]: DeleteKind -> KIND

File: pkg/cmd/subcommand/demo.go [hunk 30]
```diff
@@ -0,0 +1,490 @@
...
+	}
+
+	return nil
+}
+
+func (i *Demo) RunOpenFunctionDemo(ctx context.Context, cl *k8s.Clientset, operator *common.Operator) error {
+	ctx, done := context.WithCancel(ctx)
+	defer done()
+
+	ti := util.NewTaskInformer("OpenFunctionDemo")
```
Comment (tpiperatgod) [inline pkg/cmd/subcommand/demo.go]: OpenFunctionDemo -> DEMO

File: pkg/cmd/util/print.go [hunk 31]
```diff
@@ -212,3 +212,10 @@ func (ti *TaskInformer) TipsOnUsingKnative() {
 		"  --type merge --patch '{\"data\":{\"1.2.3.4.sslip.io\":\"\"}}'")
 	fmt.Println()
 }
+
+func (ti *TaskInformer) TipsOnOpenfunctionDemo(Endpoint string) {
+	fmt.Println(YellowItalic("Now we have configured the appropriate parameters for you, "+
+		"You can use this address to access related functions : \n"), Endpoint)
+	fmt.Println(YellowItalic("\n We now use the curl command to access the address. The following information was returned:"))
```
Comment (tpiperatgod) [inline pkg/cmd/util/print.go]: ->
```go
	fmt.Println(YellowItalic("Now we have configured the appropriate parameters for you, "+
		"You can use this address to access related functions : \n"), Endpoint)
        fmt.Println()
	fmt.Println(YellowItalic("We now use the curl command to access the address. The following information was returned:"))
```

File: testdata/demo/v0.4.0/function-sample-serving-only.yaml [hunk 32]
```diff
@@ -0,0 +1,15 @@
+apiVersion: core.openfunction.io/v1alpha2
+kind: Function
+metadata:
+  name: function-sample
+spec:
+  version: "v1.0.0"
+  image: "openfunctiondev/sample-go-func:latest"
```
Comment (tpiperatgod) [inline testdata/demo/v0.4.0/function-sample-serving-only.yaml]: image: "openfunctiondev/sample-go-func:latest"
->
image: "openfunctiondev/sample-go-func:v0.4.0"

**Review by tpiperatgod — CHANGES_REQUESTED** [review]

**Review by tpiperatgod — APPROVED** [review]

File: pkg/cmd/subcommand/demo.go [hunk 33]
```diff
@@ -0,0 +1,478 @@
...
+`,
+		Example: "ofn demo",
+		Run: func(cmd *cobra.Command, args []string) {
+			util.CheckErr(i.RunKind(cl, cmd))
+		},
+	}
+
+	cmd.Flags().BoolVar(&i.Verbose, "verbose", false, "Show verbose information.")
+	cmd.Flags().BoolVar(&i.RegionCN, "region-cn", false, "For users who have limited access to gcr.io or github.com.")
+	cmd.Flags().BoolVar(&i.AutoPrune, "auto-prune", true, "Used to specify the version of OpenFunction to be uninstalled.")
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/demo.go]: "Used to specify the version of OpenFunction to be uninstalled." is not correct for auto-prune

File: pkg/cmd/subcommand/demo.go [hunk 34]
```diff
@@ -0,0 +1,485 @@
...
+`,
+		Example: "ofn demo",
+		Run: func(cmd *cobra.Command, args []string) {
+			util.CheckErr(i.RunKind(cl, cmd))
+		},
+	}
+
+	cmd.Flags().BoolVar(&i.Verbose, "verbose", false, "Show verbose information.")
+	cmd.Flags().BoolVar(&i.RegionCN, "region-cn", false, "For users who have limited access to gcr.io or github.com.")
+	cmd.Flags().BoolVar(&i.AutoPrune, "auto-prune", true, "Automatic cleaning of the current environment.")
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/demo.go]: Automatically clean up the demo environment.

**Review by benjaminhuo — APPROVED** [review]

**Pull request merged**